### PR TITLE
PERF: make RangeIndex iterate over ._range

### PR DESCRIPTION
**Pull request by topper-123**
Minor performance issue.

By adding a custom ``__iter__`` method to ``RangeIndex``, we partly avoid needing to create/cache the expensive ``_data`` attribute and partly it's just faster to iterate over a ``range`` than a ``ndarray``:

```python
>>> idx = pd.RangeIndex(100_000)
>>> %%timeit
... for _ in idx:
...     pass
10.9 ms ± 74.7 µs per loop  # master
6.11 ms ± 48.8 µs per loop  # this PR
>>> "_data" in idx._cache
True  # master
False  # this PR
```

xref #35432, #26565.

### CODE REVIEW

File: pandas/tests/indexes/ranges/test_range.py [hunk 1]
```diff
@@ -167,6 +167,10 @@ def test_cache(self):
         idx.any()
         assert idx._cache == {}
 
+        for _ in idx:
+            ...
```
Comment (jbrockmendel) [inline pandas/tests/indexes/ranges/test_range.py:171]: i think "pass" would be more idiomatic here

Comment (topper-123) [reply pandas/tests/indexes/ranges/test_range.py]: Ok

**Comment by jreback** [comment]
cool, can you add an asv which hit this (or do we have enough coverage)?

**Comment by topper-123** [comment]
Ok, i’ve added ASVs.

**Comment by jreback** [comment]
thanks @topper-123

**Pull request merged**